### PR TITLE
[INLONG-7959][Sort] Dynamic schema evolution support delete and update columns when sink to Iceberg

### PR DESCRIPTION
**Pull request by hejiay**
I accidentally deleted the original PR https://github.com/apache/inlong/pull/8095 by mistake, so I resubmitted it.For details, see this PR https://github.com/apache/inlong/pull/8095

### CODE REVIEW

**Comment by hejiay** [comment]
@dockerzhang

**Review by EMsnap — APPROVED** [review]
Looks good for me, @lordcheng10 PLAL as well if it doesn't affect your current work

**Review by dockerzhang — APPROVED** [review]

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java [hunk 1]
```diff
@@ -52,49 +54,85 @@ public class SchemaChangeUtils {
...
-                ni++;
-            } else {
-                NestedField newField = newSchema.findField(newFields.get(ni));
+
+        // step0: judge whether unknown change
+        // just diff two different schema can not distinguish（add + delete) vs modify
+        // Example first [a, b, c] -> then delete c [a, b] -> add d [a, b, d], currently it is only judged as unknown
+        // change.
+        // In next version,we will judge it is [delete and add] or rename by using information extracted from DDL
+        if (!colsToDelete.isEmpty() && !colsToAdd.isEmpty()) {
```
Comment (thexiay) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java:71]: whether to add a reorder column check?

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java [hunk 2]
```diff
@@ -52,49 +54,85 @@ public class SchemaChangeUtils {
...
+            if (colsToDelete.contains(colName)) {
+                tableChanges.add(
+                        new TableChange.DeleteColumn(
+                                new String[]{colName}));
+            }
         }
 
+        // step3: judge whether add column
+        if (!colsToAdd.isEmpty()) {
+            for (int i = 0; i < newFields.size(); i++) {
```
[the remark this answers is not in the document]
Comment (hejiay) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java:103]: In order to get the position of the added field

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java (hunk 1 above)
Comment (hejiay) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java:71]: Excuse me, could you elaborate more?

Comment (thexiay) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java:71]: [a,b,c]->[a,c,b]
it'a possible to happen，so you should consider it as an abnormal case or just ignore it.It should clearly described in the method comment. And you should add some ut to validate the input and output.

Comment (hejiay) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java:71]: Thanks,I get it and I will perfect it

**Review by gong — APPROVED** [review]

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java (hunk 1 above)
Comment (hejiay) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java:70]: I added judge field position change, please check again

**Review by thexiay — APPROVED** [review]

**Comment by hejiay** [comment]
The idea of ​​this PR is mainly provided by @sparklezzz, I am mainly responsible for practice and testing, thanks to @sparklezzz

**Comment by lordcheng10** [comment]
How is column reorder handled? Throw unsupport exception?  @hejiay @thexiay

**Pull request merged**